src/datasets/arrow_dataset.py

```diff
@@ -228,171 +239,201 @@ def _get_output_signature(dataset: "Dataset", collate_fn: Callable, collate_fn_a
+            )
+        # endregion
+
+        # region Argument and default value handling
+        if collate_fn is None:
+            # Set a very simple default collator that just stacks things together
+            collate_fn = minimal_tf_collate_fn
         if collate_fn_args is None:
             collate_fn_args = {}
-
+        if label_cols and not columns:
+            raise ValueError("Cannot specify label_cols without specifying columns!")
         if label_cols is None:
             label_cols = []
         elif isinstance(label_cols, str):
             label_cols = [label_cols]
-        elif len(set(label_cols)) < len(label_cols):
+        if len(set(label_cols)) < len(label_cols):
             raise ValueError("List of label_cols contains duplicates.")
-        if not columns:
-            raise ValueError("Need to specify at least one column.")
-        elif isinstance(columns, str):
-            columns = [columns]
-        elif len(set(columns)) < len(columns):
-            raise ValueError("List of columns contains duplicates.")
-        if label_cols is not None:
-            cols_to_retain = columns + label_cols
+        if columns:
+            if isinstance(columns, str):
+                columns = [columns]
+            if len(set(columns)) < len(columns):
+                raise ValueError("List of columns contains duplicates.")
+            cols_to_retain = list(set(columns + label_cols))
         else:
-            cols_to_retain = columns
-        if "label" in cols_to_retain or "labels" in cols_to_retain or "label_ids" in cols_to_retain:
-            cols_to_retain += ["labels", "label", "label_ids"]  # Don't accidentally drop any labels with other names!
-        cols_to_retain = list(set(cols_to_retain))  # Remove any duplicates
-
+            cols_to_retain = None  # Indicates keeping all columns
+            columns = []
         if drop_remainder is None:
             # We assume that if you're shuffling it's the train set, so we drop the remainder unless told not to
             drop_remainder = shuffle
+        # endregion
+
+        # region Set dataset format and drop unwanted
+
+        if self.format["type"] != "custom":
+            dataset = self.with_format("numpy")
+        else:
+            dataset = self
+        if cols_to_retain is not None:
+            # Following the logic in `transformers.Trainer`, we do not drop `label_ids` or `label` even if they
+            # are not in the list of requested columns, because the collator may rename them
+            # This might work better if moved to a method attached to our transformers Model objects, but doing so
+            # could break backward compatibility
+            unwanted_columns = [
+                col
+                for col in self.features.keys()
+                if col not in columns and col not in label_cols and col not in ("label_ids", "label")
+            ]
+            dataset = dataset.remove_columns(unwanted_columns)
```

Comment:
One thing I'm still not sure about is that we special-case "label_ids" and "label", because those columns are frequently renamed to "labels" in `transformers` and we don't want to drop them. In PyTorch this special-casing occurs too, but it happens inside `Trainer`, so it doesn't mix things up between the `transformers` and `datasets` libraries. We might have to leave it here for now until I can make the 'magic method' on our `transformers` models!

Reply:
The standard workflow for `transformers` is that the column in the dataset is called `label`, but the argument to the model is called `labels`, and the renaming is done by the `collate_fn`. Therefore, we must be careful not to drop columns called `label` or `label_ids`, even if the user doesn't request them, because the `collate_fn` might need them.

In PyTorch columns are dropped [by the Trainer](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L610), and it makes sure not to drop these. However, in Tensorflow, columns are dropped by the `to_tf_dataset()` method, and therefore this code needs to be in there.

I think a good solution would be to make a method in `transformers` that calls `to_tf_dataset()` and converts datasets for training, and then we could move the `label` and `label_ids` special casing in there. However, until we do that, we'll need to keep the special casing in `to_tf_dataset()` or all our examples will break!